asf_heat_pump_suitability/pipeline/run/add_features.py

```diff
@@ -0,0 +1,107 @@
+
+    Returns:
+        argparse.Namespace: populated `Namespace`
+    """
+    parser = argparse.ArgumentParser()
+
+    # TODO this is a placeholder and likely to change as the script develops
+    parser.add_argument(
+        "--uprns",
+        help="Path to domestic UPRN dataset with X and Y coordinates in parquet.",
+        type=str,
+        required=True,
+    )
+
+    return parser.parse_args()
+
+
+if __name__ == "__main__":
+    import os
+    import polars as pl
+    import geopandas as gpd
+
+    from asf_heat_pump_suitability.utils import save_utils
+    from asf_heat_pump_suitability.getters import load_tree_input
+    from asf_heat_pump_suitability.pipeline.impute import property_type
+    from asf_heat_pump_suitability.pipeline.transform import uprns, outdoor_space
+
+    args = parse_arguments()
+
+    # Load UPRN data
+    print(f"Loading domestic UPRNs from: {args.uprns}")
+    uprns_df = pl.read_parquet(
+        args.uprns, columns=["UPRN", "X_COORDINATE", "Y_COORDINATE"]
+    )
+
+    # Get geopoints of UPRNs
+    uprns_gdf = uprns.generate_gdf_uprn_coords(df=uprns_df)
+
+    # ------------------------ #
+    # IMPUTE PROPERTY TYPE FLAT
+    # Create boolean column called `property_type_flat` to identify flats
+    flat_uprns = property_type.impute_set_flat_properties(uprns_gdf=uprns_gdf)
+    features_df = uprns_df.with_columns(
+        pl.col("UPRN").is_in(flat_uprns).alias("property_type_flat")
+    )
+
+    # ------------------------ #
+    # ESTIMATE OUTDOOR SPACE
+    # TODO scale beyond Plymouth
+    print("Loading land registry data...")
+    land_parcels_gdf = gpd.read_file(
+        "s3://asf-heat-pump-suitability/local_heat_planning/plymouth_inputs/Plymouth_Land_Registry_Cadastral_Parcels.gml"
+    )
+    building_footprints_gdf = load_tree_input.load_gdf_os_openmap_local_layer(
+        layer="building", grid_squares="SX"
+    )
+
+    # Get intersection of building footprint polygons and land polygons
+    intersection_gdf = outdoor_space.generate_gdf_building_intersections(
+        land_parcels_gdf=land_parcels_gdf,
+        building_footprints_gdf=building_footprints_gdf,
+    )
+
+    # Get outdoor space
+    outdoor_space_gdf = outdoor_space.generate_gdf_outdoor_space(
+        intersection_gdf, land_parcels_gdf
+    )
+    uprns_space_df = outdoor_space.sjoin_df_uprn_to_outdoor_space(
+        uprns_gdf, outdoor_space_gdf
+    )
+    uprns_space_df = outdoor_space.deduplicate_df_outdoor_space(uprns_space_df)
+
+    # Join outdoor space estimates to dataframe
+    features_df = features_df.join(
+        uprns_space_df.select(
+            [
+                "UPRN",
+                "NATIONALCADASTRALREFERENCE",
+                "max_contiguous_outdoor_space_area_m2",
+                "total_outdoor_space_area_m2",
+            ]
+        ),
+        how="left",
+        on="UPRN",
+    )
+
+    # ------------------------ #
+    # SAVE OUTPUTS
+    save_utils.save_to_s3(
```

Comment:
When I last checked this file had more lines than UPRNs, so there might be somewhere where we are not deduplicating?